app/services/experiment_control_service.rb

```diff
@@ -0,0 +1,94 @@
+class InvalidExperiment < StandardError; end
+
+class ExperimentControlService
+  class << self
+    def start_current_patient_experiment(name, days_til_start, days_til_end)
+      raise ArgumentError, "Start date must be before end date" if days_til_end < days_til_start
+
+      existing_experiment = Experimentation::Experiment.where(experiment_type: "current_patient_reminder", state: ["selecting", "live"])
+      raise InvalidExperiment, "A current patient experiment is already in progress" if existing_experiment.any?
+
+      experiment = Experimentation::Experiment.find_by!(name: name, experiment_type: "current_patient_reminder")
+      experiment_start = days_til_start.days.from_now.beginning_of_day
+      experiment_end = days_til_end.days.from_now.end_of_day
+
+      experiment.state_selecting!
+
+      experiment_patients = patient_pool
+        .joins(:appointments)
+        .where("appointments.status = ?", "scheduled")
+        .where("appointments.scheduled_date BETWEEN ? AND ?", experiment_start, experiment_end)
+        .order(Arel.sql("random()"))
+
+      experiment_patients.each do |patient|
+        group = experiment.group_for(patient.id)
+        appointment = patient.appointments
+          .where(status: "scheduled")
+          .where("appointments.scheduled_date BETWEEN ? AND ?", experiment_start, experiment_end).first
+        schedule_reminders(patient, appointment, experiment, group, appointment.scheduled_date)
+        Experimentation::TreatmentGroupMembership.create!(treatment_group: group, patient: patient)
+      end
+
+      experiment.update!(state: "live", start_date: experiment_start, end_date: experiment_end)
+    end
+
+    def start_stale_patient_experiment(name, total_days)
+      experiment = Experimentation::Experiment.find_by!(name: name, experiment_type: "stale_patient_reminder")
+      date = Date.current
+      eligibility_start = (date - 365.days).beginning_of_day
+      eligibility_end = (date - 35.days).end_of_day
+
+      experiment.state_selecting!
+
+      eligible_patients = patient_pool
+        .joins(:encounters)
+        .where("encounters.device_created_at BETWEEN ? AND ?", eligibility_start, eligibility_end)
+        .where("NOT EXISTS (SELECT 1 FROM encounters WHERE encounters.patient_id = patients.id AND
+                encounters.device_created_at > ?)", eligibility_end)
+        .left_joins(:appointments)
+        .where("NOT EXISTS (SELECT 1 FROM appointments WHERE appointments.patient_id = patients.id AND
+                appointments.scheduled_date >= ?)", date)
+        .order(Arel.sql("random()"))
+        .to_a
```

Comment:
We should definitely not be doing `to_a` on a large relation of patients like this.  Lets discuss this during a pair session tomorrow.

Reply:
Agreed. Performance was something I wanted your input on for this script in general. I'm sure we can just select ranges or something rather than popping.